conjur/init_controller.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+
+"""
+InitController module
+
+This module is the controller that facilitates all init actions
+required to successfully configure the conjurrc
+"""
+
+# Builtin
+import logging
+import os
+import sys
+
+from urllib.parse import urlparse
+
+import OpenSSL
+
+class InitController:
+    """
+    InitController
+
+    This class represents the Presentation Layer for the init command
+    """
+    DEFAULT_CONFIG_FILE = os.path.expanduser(os.path.join('~', '.conjurrc'))
+
+    force_overwrite = False
+    init_input_obj = None
+    init_logic = None
+
+    def __init__(self, init_input_obj, init_logic, force):
+        self.init_input_obj = init_input_obj
+        self.init_logic = init_logic
+        self.force_overwrite = force
+
+    def call(self):
+        """
+        Method that facilitates all method calls in this class
+        """
+        fetched_certificate = self.get_server_certificate(self.init_logic,
+                                                          self.init_input_obj)
+        self.get_host_info(self.init_input_obj)
+        self.write_certificate(self.init_logic,
+                               self.init_input_obj,
+                               self.init_input_obj.cert_file,
+                               fetched_certificate)
+        self.init_logic.write_conjurrc(self.DEFAULT_CONFIG_FILE,
+                                       self.init_input_obj,
+                                       self.force_overwrite)
+
+    # pylint: disable=line-too-long
+    @staticmethod
+    def get_server_certificate(init_logic, input_obj):
+        """
+        Method to get the certificate from the Conjur endpoint detailed by the user
+        """
+        url = input_obj.appliance_url if input_obj.appliance_url else input("Enter the URL of your Conjur service: ")
+        input_obj.appliance_url = ''.join(url)
+
+        url = urlparse(input_obj.appliance_url)
+
+        # Fetch certificate information if connection is expected over TLS
+        certificate_path = input_obj.cert_file if input_obj.cert_file else None
+        if certificate_path is not None:
+            input_obj.cert_file = ''.join(certificate_path)
+
+        if certificate_path is None and url.scheme == "https":
```

Comment:
This logic to fetch the cert should be in it's own method/file

Reply:
the actual logic is. It is in `init_logic.get_certificate` on line 73. The controller is handling just the method calls. In the actual business logic class, I fetch connect to the conjur socket and fetch the cert

Reply:
Sure but the certificate fetching is such a separate concern from this class that it can (and probably should be) moved out into its own class.

Reply:
Are we talking about moving this line
```
            fingerprint, fetched_certificate = init_logic.get_certificate(url.hostname,
                                                                          url.port, sock,
                                                                          OpenSSL.crypto.dump_certificate)
```
to a separate class or the actual logic _inside_ the `get_certificate()` function?
if it's the actual logic of opening the socket then sure! otherwise seems like a lot of work just for a single line of code

Reply:
separated out to a service